lib/dependencies.js

```diff
@@ -0,0 +1,384 @@
+    let oldRootDir = dependencyData.rootDir;
+    newRootDir = newRootDir || oldRootDir;
+    let dependencies = new Dependencies(newRootDir);
+    let files = Object.keys(dependencyData.dependencies);
+    for (let file of files) {
+      let deps = dependencyData.dependencies[file];
+      if (newRootDir) {
+        for (let i = 0; i < deps.length; i++) {
+          let dep = deps[i];
+          if (dep.startsWith(oldRootDir+path.sep)) {
+            deps[i] = dep.replace(oldRootDir, newRootDir);
+          }
+        }
+      }
+      dependencies.setDependencies(file, deps);
+    }
+    /** @type {Map<string, FSTree>} */
+    let fsTrees = new Map();
+    for (let fsTreeData of dependencyData.fsTrees) {
+      /** @type {Array<Entry | HashEntry>} */
+      let entries = [];
+      for (let entry of fsTreeData.entries) {
+        if (entry.type === "stat") {
+          entries.push(new Entry(entry.relativePath, entry.size, entry.mtime, entry.mode));
+        } else {
+          entries.push(new HashEntry(entry.relativePath, entry.hash));
+        }
+      }
+      /** @type {FSTree | FSHashTree}  */
+      let fsTree;
+      /** @type {string} */
+      let treeRoot;
+      if (fsTreeData.fsRoot === oldRootDir) {
+        treeRoot = newRootDir;
+        fsTree = FSHashTree.fromHashEntries(entries, { sortAndExpand: true });
+      } else {
+        treeRoot = fsTreeData.fsRoot;
+        fsTree = FSTree.fromEntries(entries, { sortAndExpand: true });
+      }
+      fsTrees.set(treeRoot, fsTree);
+    }
+    dependencies.seal();
+    dependencies.fsTrees = fsTrees;
+    return dependencies;
+  }
+};
+
+/**
+ * Get an FSTree that uses content hashing information to compare files to
+ * see if they have changed.
+ *
+ * @param fsRoot {string}
+ * @param dependencies {Set<string>}
+ * @return {FSHashTree}
+ */
+function getHashTree(fsRoot, dependencies) {
+  /** @type {Array<HashEntry>} */
+  let entries = [];
+  for (let dependency of dependencies) {
+    let fullPath = path.join(fsRoot, dependency);
+    try {
+      // it would be good if we could cache this and share it with
+      // the read that accompanies `processString()` (if any).
+      let contents = fs.readFileSync(fullPath, "utf8");
+      let hash = md5sum(contents);
+      entries.push(new HashEntry(dependency, hash));
+    } catch(e) {
+      entries.push(new HashEntry(dependency, ""));
+    }
+  }
+  return FSHashTree.fromHashEntries(entries);
+}
+
+/**
+ * Get an FSTree that uses fs.stat information to compare files to see
+ * if they have changed.
+ *
+ * @param fsRoot {string}
+ * @param dependencies {Set<string>}
+ */
+function getStatTree(fsRoot, dependencies) {
+  /** @type {Array<Entry>} */
+  let entries = [];
+  for (let dependency of dependencies) {
+    let fullPath = path.join(fsRoot, dependency);
+    try {
+      // TODO: Share a cache of stat results across all persistent filters.
+      let stats = fs.statSync(fullPath);
```

Comment:
Can we reduce the openess of this try/catch? I have trolled myself too often which such try/catches, specifically as overtime code bit-rots, this pattern can "gracefully" absorb more then intended, resulting in silent failures.

I believe the following should do the trick.

```js
let stat;

try {
  stat = fs.statSync(fullPath);
} catch (e) {
  if (e !== null && typeof e === 'object' && e.code === 'ENOENT') {
    // gracefully recover from the stat pointing to non-existent FD, by providing a "null" descriptor
    stat = { size: 0, mtime: 0, mode: 0 };
  } else {
    // rethrow unknown errors, as we really don't know if it is safe to recover
    throw e;
  }
}

entries.push(Entry.fromStat(dependency, stat);
```